OpenDreamClient/Rendering/ClientImagesSystem.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using OpenDreamShared.Dream;
+using OpenDreamShared.Rendering;
+
+namespace OpenDreamClient.Rendering {
+    sealed class ClientImagesSystem : SharedClientImagesSystem {
+        private readonly Dictionary<IconAppearance, List<uint>> TurfClientImages = new();
+        private readonly Dictionary<EntityUid, List<uint>> AMClientImages = new();
+        private readonly Dictionary<uint, DreamIcon> _idToIcon = new();
+        [Dependency] private readonly ClientAppearanceSystem clientAppearanceSystem = default!;
+        [Dependency] private IEntityManager _entityManager = default!;
+
+        public override void Initialize() {
+            SubscribeNetworkEvent<AddClientImageEvent>(OnAddClientImage);
+            SubscribeNetworkEvent<RemoveClientImageEvent>(OnRemoveClientImage);
+        }
+
+        public override void Shutdown() {
+            TurfClientImages.Clear();
+            AMClientImages.Clear();
+        }
+
+        public bool TryGetClientImages(EntityUid entity, IconAppearance appearance, [NotNullWhen(true)] out List<DreamIcon>? result){
+            result = null;
+            List<uint>? resultIDs;
+            if(entity == EntityUid.Invalid) {
+                if(!TurfClientImages.TryGetValue(appearance, out resultIDs))
+                    return false;
+            }
+            else {
+                if(!AMClientImages.TryGetValue(entity, out resultIDs))
+                    return false;
+            }
+            result = new List<DreamIcon>();
+            foreach(uint distinctID in resultIDs.Distinct())
```

Comment:
Are these IDs not inherently distinct? `.Distinct()` can be rather slow.

Reply:
They are not. Because Turfs share appearances, adding the same client image to multiple turfs results in one appearance with that client image's id added multiple times. Since we only want to draw it once, these must be made distinct.
Additionally this acts to keep track of the number of times that client image is used, since one of the turfs could have it removed without it affecting the others.

Reply:
Turfs don't share appearances anymore, removed the `Distinct()` call